sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/AutoConfigurationCustomizer.java

```diff
@@ -62,6 +63,23 @@ AutoConfigurationCustomizer addSpanExporterCustomizer(
       BiFunction<? super SpanExporter, ConfigProperties, ? extends SpanExporter>
           exporterCustomizer);
 
+  /**
+   * Adds a {@link BiFunction} to invoke with the default autoconfigured {@link
+   * io.opentelemetry.sdk.trace.SpanProcessor}s responsible for invoking registered {@link
+   * SpanExporter}s. The return value of the {@link BiFunction} will replace the passed-in argument.
+   * In contrast to {@link #addSpanExporterCustomizer(BiFunction)} this allows modifications to
+   * happen before batching occurs. As a result, it is possible to efficiently filter spans, add
+   * artificial spans or delay spans for enhancing them with external, delayed data. The provided
+   * function will be invoked at most once per customized SDK instance.
+   *
+   * <p>Multiple calls will execute the customizers in order.
+   */
+  default AutoConfigurationCustomizer addBatchSpanProcessorCustomizer(
```

Comment:
```suggestion
  default AutoConfigurationCustomizer addSpanProcessorCustomizer(
```

Reply:
will this customizer be applied to all span processors, or only to the batch span processor?

Reply:
As written in this PR currently it applies to the result of `SpanProcessor.composite`, which may be a batch processor or a composite processor depending on whether one or multiple exporters are configured.

I don't see why we would want to limit it to batch processor. Right now, there is no way for other processors to be automatically configured besides simple and batch processor, but with file configuration that will change. If we end up providing the ability to customize the components (notably, I don't think we should allow customizing components), then there will be other potential processors besides simple and batch.

Reply:
Why did I go for customizing the result of `SpanProcessor.composite` instead of the individual span processors?
Consider the use-case from my initial comment:
> Delaying spans
E.g. in order to add profiling statistics to spans as attributes. Those statistics are likely not immediately available on span end.

The idea here was, that even if multiple span-exporting `SpanProcessors` are registered, the "delaying" done by the customizer and resulting buffering only needs to be done once per span.
If you instead wrap each individual `SpanProcessor`, you'll need to do the buffering for each one: E.g. if you have both a `SimpleSpanProcessor` and a `BatchSpanProcessor` registered, you would end up with double the memory consumed for buffering and doing the correlation with the data you are waiting for two times.

In production environments I would however expect that you only have a single `BatchSpanProcessor` and no `SimpleSpanProcessor`, so the proposal to wrap individual `SpanProcessors` is fine for me. I'll then just remove the following guarantee from the javadoc:
```
The provided function will be invoked at most once per customized SDK instance.
```

> Right now, there is no way for other processors to be automatically configured besides simple and batch processor, but with file configuration that will change.

I'm curious, do you have any link to an Issue/PR/something to read up on that feature?

I originally named the method `addSpanExportingProcessorCustomizer` to explicitly state that the customizer would only be invoked for `SpanProcessor`s responsible for invoking `SpanExporter`s. The reason is that I did not want to wrap other `SpanProcessors`, e.g. ones responsible for lifting baggage into span-attributes.

I guess with your proposal this still can be done by doing an `instanceof BatchSpanProcessor` check in the customizer, though this feels a little hacky.

Reply:
>> Right now, there is no way for other processors to be automatically configured besides simple and batch processor, but with file configuration that will change.

> I'm curious, do you have any link to an Issue/PR/something to read up on that feature?

Here are some resources:
- The original [OTEP](https://github.com/open-telemetry/oteps/blob/main/text/0225-configuration.md)
- The relevant portion of the [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/file-configuration.md)
- The [opentelemetry-configuration](https://github.com/open-telemetry/opentelemetry-configuration) repository defining the file schema
- The Configuration SIG meets every other monday at 8am PT. Details [here](https://github.com/open-telemetry/community#specification-sigs).

> I guess with your proposal this still can be done by doing an instanceof BatchSpanProcessor check in the customizer, though this feels a little hacky.

This is the expectation right now with all the customizers (i.e. `addSamplerCustomizer`, `add{Signal}ExporterCustomizer`, `addPropagatorCustomizer`), so I think its the right decision.

Another feature of the customizers is they give you the ability to fully replace the component in addition to the ability to wrap it with a delegating class. I'm imagine you could use this with processor customizer for an additional use case: Suppose you don't want to use the batch processor at all, but you know that autoconfigured exporters always get paired with a batch processor. If we add an additional `getExporter()` method to `BatchSpanProcessor`, then you could use a span processor customizer to replace the batch processor with another, while retaining the autoconfigured exporter. Something like:
```
    autoConfiguration.addSpanProcessorCustomizer(
        (spanProcessor, configProperties) -> {
          if (spanProcessor instanceof BatchSpanProcessor) {
            SpanExporter exporter = ((BatchSpanProcessor) spanProcessor).getExporter();
            return new MyCustomProcessor(exporter);
          }
          return spanProcessor;
        });
```
Pretty slick, IMO.